book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,220 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+V5 transactions are described by the protocol in the second table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus).
+
+We need the raw representation to Serialize/Deserialize the transaction however Zebra uses its own representation internally.
+
+V5 transactions are the only ones that will support orchard transactions with `Orchard` data types.
+
+V4 and V5 transactions both support sapling data but they are implemented differently. By this reason we need to split sapling data types into V4 and V5.
+
+The order of some of the fields changed from V4 to V5. For example the `lock_time` and `expiry_height` were moved above the transparent inputs and outputs.
+
+All of the changes proposed in this document are only to the `zebra-chain` crate.
+
+To highlight changes all document comments from the code snippets in the [reference section](#reference-level-explanation) were removed.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We start by looking how a V4 (already implemented) transaction is represented in Zebra.
+
+This transaction version is specified by the protocol in the first table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus).
+
+
+```
+V4 {
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    value_balance: Amount,
+    joinsplit_data: Option<JoinSplitData<Groth16Proof>>,
+    shielded_data: Option<ShieldedData>,
+}
+```
+
+Currently the `ShieldedData` type is defined in `zebra-chain/src/transaction/shielded_data.rs` as follows:
+
+```
+pub struct ShieldedData {
+    pub first: Either<Spend, Output>,
+    pub rest_spends: Vec<Spend>,
+    pub rest_outputs: Vec<Output>,
+    pub binding_sig: Signature<Binding>,
+}
+```
+
+We know by protocol (2nd table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus)) that V5 transactions will support sapling data however we also know by protocol that spends ([Spend Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#spendencodingandconsensus), See †) and outputs ([Output Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#outputencodingandconsensus), See †) fields change from V4 to V5.
+
+Here we have the proposed changes for V4 transactions:
+
+```
+V4 {
+    ...
+    sapling_value_balance: Amount,
+    ...
+    sapling_shielded_data: Option<SaplingShieldedDataV4>,
+}
+```
+
+`ShieldedData` is currently defined and implemented in `src/transaction/shielded_data.rs`. We propose to change `ShieldedData` into `SaplingShieldedDataV4` and to add `SaplingShieldedDataV5` into the same file.
```

Comment:
Let's also rename the file to `src/transaction/sapling_shielded_data.rs`, for consistency with `src/orchard_shielded_data.rs`.

Reply:
https://github.com/ZcashFoundation/zebra/pull/1886/commits/f0aff2d9731d6613e5ae4b5ddc71b5569cbbcd9a